libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/ManeuverProcessor.kt

```diff
@@ -3,92 +3,290 @@ package com.mapbox.navigation.ui.maneuver
             }
-            is ManeuverAction.GetAllBannerInstructionsAfterStep -> {
-                processBannerInstructionsAfterStep(action.routeProgress, action.bannerInstructions)
+                ?: ManeuverResult.GetManeuverListWithProgress.Failure(
+                    "$currentBanner cannot be null"
+                )
+        } catch (exception: Exception) {
+            ManeuverResult.GetManeuverListWithProgress.Failure(exception.message)
+        }
+    }
+
+    private fun createManeuverList(
+        route: DirectionsRoute,
+        maneuverState: ManeuverState,
+        distanceFormatter: DistanceFormatter
+    ) {
+        ifNonNull(route.legs()) { routeLegs ->
+            routeLegs.forEach { routeLeg ->
+                ifNonNull(routeLeg?.steps()) { steps ->
+                    val stepList = mutableListOf<StepIndexToManeuvers>()
+                    for (stepIndex in 0..steps.lastIndex) {
+                        steps[stepIndex].bannerInstructions()?.let { bannerInstruction ->
+                            val maneuverList = mutableListOf<Maneuver>()
+                            bannerInstruction.forEach { banner ->
+                                maneuverList.add(transformToManeuver(banner, distanceFormatter))
+                            }
+                            val stepIndexToManeuvers = StepIndexToManeuvers(
+                                stepIndex,
+                                maneuverList
+                            )
+                            stepList.add(stepIndexToManeuvers)
+                        } ?: throw RuntimeException("LegStep should have valid banner instructions")
+                    }
+                    maneuverState.allManeuvers.add(LegToManeuvers(routeLeg, stepList))
+                } ?: throw RuntimeException("RouteLeg should have valid steps")
             }
-            is ManeuverAction.GetAllManeuvers -> {
-                processManeuverList(action.bannerInstructions)
+        } ?: throw RuntimeException("Route should have valid legs")
+        if (maneuverState.allManeuvers.isEmpty()) {
+            throw RuntimeException("Maneuver list could not be created")
+        }
+    }
+
+    private fun readManeuverListWith(
+        list: List<LegToManeuvers>,
+        routeLeg: RouteLeg? = null
+    ): List<Maneuver> {
+        if (list.isEmpty()) {
+            throw RuntimeException("$list cannot be empty")
+        }
+        val maneuverList = mutableListOf<Maneuver>()
+        when (routeLeg == null) {
+            true -> {
+                list[0].stepIndexToManeuvers.forEach { stepIndexToManeuver ->
+                    maneuverList.addAll(stepIndexToManeuver.maneuverList)
+                }
             }
-            is ManeuverAction.GetManeuver -> {
-                processCurrentManeuver(action.bannerInstruction)
+            else -> {
+                list.find { item -> item.routeLeg == routeLeg }?.let { legToManeuver ->
```

Comment:
Yes we still need this comparison, because we are not yet using `legIndex`. So far we haven't seen any performance issues with this comparison. If we do in the future we would introduce `legIndex` instead of `routeLeg`.

cc @Guardiola31337 @LukasPaczos